File: backend/app/schemas/evidence_seeker.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field

from app.models.evidence_seeker import FactCheckPublicationMode

from .evidence_seeker_settings import (
    ConfigurationStateLiteral,
    SetupModeLiteral,
)


class EvidenceSeekerBase(BaseModel):
    """Base schema for EvidenceSeeker"""

    title: str = Field(..., min_length=1, max_length=100)
    description: str | None = Field(None, max_length=500)
    is_public: bool = Field(default=False, alias="isPublic")
    fact_check_publication_mode: FactCheckPublicationMode = Field(
        default=FactCheckPublicationMode.AUTOPUBLISH,
        alias="factCheckPublicationMode",
    )
    language: str | None = Field(
        default=None,
        max_length=16,
        description="Language code used for preprocessing (e.g. EN, DE)",
    )

    class Config:
        populate_by_name = True


class InitialConfiguration(BaseModel):
    """Initial configuration payload for wizard onboarding."""

    api_key_name: str = Field(..., min_length=1, max_length=100, alias="apiKeyName")
    api_key_value: str = Field(..., min_length=10, alias="apiKeyValue")
    bill_to: str | None = Field(None, alias="billTo", max_length=100)
    setup_mode: SetupModeLiteral = Field(
        default="SIMPLE", alias="setupMode", description="Setup mode selection"
    )

    class Config:
        populate_by_name = True


class EvidenceSeekerCreate(EvidenceSeekerBase):
    """Schema for creating EvidenceSeeker"""

    initial_configuration: InitialConfiguration | None = Field(
        default=None, alias="initialConfiguration"
    )


class EvidenceSeekerUpdate(BaseModel):
    """Schema for updating EvidenceSeeker"""

    title: str | None = Field(None, min_length=1, max_length=100)
    description: str | None = Field(None, max_length=500)
    is_public: bool | None = Field(None, alias="isPublic")
    fact_check_publication_mode: FactCheckPublicationMode | None = Field(
        None, alias="factCheckPublicationMode"
    )
    language: str | None = Field(
        default=None,
        max_length=16,
        description="Language code used for preprocessing (e.g. EN, DE)",
    )

    class Config:
        populate_by_name = True


class EvidenceSeekerRead(EvidenceSeekerBase):
    """Schema for reading EvidenceSeeker"""

    id: int
    uuid: UUID  # External API identifier
    logo_url: str | None = Field(alias="logoUrl")
    created_by: int = Field(alias="createdBy")
    created_at: datetime = Field(alias="createdAt")
    updated_at: datetime = Field(alias="updatedAt")
    published_at: datetime | None = Field(alias="publishedAt", default=None)
    is_public: bool = Field(alias="isPublic", default=False)
    fact_check_publication_mode: FactCheckPublicationMode = Field(
        alias="factCheckPublicationMode",
        default=FactCheckPublicationMode.AUTOPUBLISH,
    )
    configuration_state: ConfigurationStateLiteral | None = Field(
        alias="configurationState", default=None
    )
    missing_requirements: list[str] = Field(
        alias="missingRequirements", default_factory=list
    )
    language: str | None = Field(default=None)
    configured_at: datetime | None = Field(alias="configuredAt", default=None)
    setup_mode: SetupModeLiteral | None = Field(alias="setupMode", default=None)
    document_skip_acknowledged: bool = Field(
        alias="documentSkipAcknowledged", default=False
    )
    onboarding_token: str | None = Field(
        alias="onboardingToken", default=None, description="Short-lived wizard token"
    )

    class Config:
        populate_by_name = True
        from_attributes = True
        by_alias = True

File: frontend/src/types/evidenceSeeker.ts
export type ConfigurationState =
  | "UNCONFIGURED"
  | "MISSING_CREDENTIALS"
  | "MISSING_DOCUMENTS"
  | "READY"
  | "ERROR";

export type SetupMode = "SIMPLE" | "EXPERT";
export type FactCheckPublicationMode = "AUTOPUBLISH" | "MANUAL";

export interface ConfigurationStatus {
  state: ConfigurationState;
  setupMode: SetupMode;
  configuredAt: string | null;
  missingRequirements: string[];
  isReady: boolean;
  documentSkipAcknowledged: boolean;
}

export interface EvidenceSeeker {
  id: number;
  uuid: string; // External API identifier
  title: string;
  description: string;
  language?: string | null;
  logoUrl: string | null;
  isPublic: boolean;
  factCheckPublicationMode: FactCheckPublicationMode;
  publishedAt?: string | null;
  createdBy: number;
  createdAt: string;
  updatedAt: string;
  configurationState?: ConfigurationState | null;
  missingRequirements?: string[];
  configuredAt?: string | null;
  setupMode?: SetupMode | null;
  documentSkipAcknowledged?: boolean;
  onboardingToken?: string | null;
}

export interface EvidenceSeekerInitialConfiguration {
  apiKeyName: string;
  apiKeyValue: string;
  billTo?: string;
  setupMode?: SetupMode;
}

export interface EvidenceSeekerCreate {
  title: string;
  description: string;
  isPublic?: boolean;
  factCheckPublicationMode?: FactCheckPublicationMode;
  language?: string;
  initialConfiguration?: EvidenceSeekerInitialConfiguration;
}

export interface EvidenceSeekerUpdate {
  title?: string;
  description?: string;
  isPublic?: boolean;
  factCheckPublicationMode?: FactCheckPublicationMode;
  language?: string | null;
}

export interface EvidenceSeekerSettings {
  evidenceSeekerId: number;
  defaultModel?: string | null;
  temperature?: number | null;
  topK?: number | null;
  rerankK?: number | null;
  maxTokens?: number | null;
  language?: string | null;
  embedBackendType: string;
  embedBaseUrl?: string | null;
  embedBillTo?: string | null;
  trustRemoteCode?: boolean | null;
  metadataFilters: Record<string, unknown>;
  pipelineOverrides?: Record<string, unknown> | null;
  huggingfaceApiKeyId?: number | null;
  lastValidatedAt?: string | null;
  updatedAt?: string | null;
  setupMode: SetupMode;
  configurationState: ConfigurationState;
  configuredAt?: string | null;
  missingRequirements: string[];
}

export interface EvidenceSeekerTestSettingsResponse {
  detail: string;
  metadataFilters: Record<string, unknown>;
}

export interface Permission {
  id: number;
  userId: number;
  evidenceSeekerId: number;
  role: "EVSE_ADMIN" | "EVSE_READER";
  createdAt: string;
}

export interface PermissionCreate {
  userId: number;
  evidenceSeekerId: number;
  role: "EVSE_ADMIN" | "EVSE_READER";
}
